### ISSUE TITLE
Fix Dockerfile

### ISSUE THREAD

**Issue by tonekk** [consul/consul#4797]
The current `Dockerfile` has some serious deficiencies.
We had huge problems to use it in production.

I would like to offer to provide a Docker setup (or maybe different ones for production and development), which is implemented according to current best practices.

**Comment by javierm** [comment]
Hi, @tonekk :smile:.

We're actually looking for someone to help us with Docker :tada:. We've got an open pull request and another one in development. Would it be possible for you to have a look at pull request #4765 (the part regarding the `.dockerignore` file) and check whether it introduces breaking changes or it can be improved in any other way?

I've just opened pull request #4798 (which depends on the previous one) so you can have a look at what we're currently working on. Maybe we could work together on that one? What do you think? :thinking:

> The goal should be to render consul/installer obsolete since modern deployment should just happen via docker.

We're still going to maintain the installer even if we improve the Docker configuration, because there are institutions who must/prefer a "classic" installation. But it would be awesome to provide both options, since AFAIK our current Docker configuration wasn't done with production environments in mind.

**Comment by muki** [comment]
Incidentally I am also currently working on Dockerizing Consul. My final mission is to deploy it to a Kubernetes cluster, but proper Dockerfiles are the necessary first step.

I can confirm that as of this morning `master` + `docker_ignore` + `docker_cleanup` build and `docker-compose up` starts a server. There's things still to polish, but I would suggest merging https://github.com/consul/consul/pull/4798 and https://github.com/consul/consul/pull/4765 as soon as possible so continued development has a stable base.

I have a few changes I'd like to PR on top of these that will hopefully simplify the Docker setup and move this issue closer to resolution.

**Comment by tonekk** [comment]
We've just dockerized docker, too.
I will create a pull request today, so maybe we can discuss some best practices

**Comment by muki** [comment]
Yes, they are redundant because those things are defined in the Dockerfile. The docker-compose stuff just overrides them (with the same values in this case).

Personally I often keep `command` there just so it's explicit, especially if the "production" command in the Dockerfile differs from the development command you'd want to run with docker-compose. But that heavily depends on the assumption that docker-compose.yaml is mostly for dev purposes (which it very often isn't).

The opinions in the previous paragraph are incredibly personal though, so I think just leaving it removed as it is right now is definitely a good enough choice.

**Comment by muki** [comment]
Awesome! I have a working deployment, but will be extremely busy for the next 14 days. I'll try and get a PR ready sooner, but if not, expect one in two to three weeks.

**Comment by javierm** [comment]
Hi, @muki @tonekk :smile:.

We're thinking about which features will be added in version 1.6, and were wondering: are you still working on this issue? If not, that's OK!

By the way, no pressure :wink:. We certainly don't want you to be stressed about including it in time for version 1.6; we're just checking what the status is.

Thanks!

**Comment by tonekk** [comment]
Hi, I could provide the Dockerfile that we use to host consul in our setup for the SVA app (mentioned in https://github.com/consul/consul/pull/4872).
Maybe it will need some tweaking to make its way into the repo, but we run it in production rn.

**Comment by javierm** [comment]
@tonekk That would be great, thanks! :smile:

**Comment by Kremer80** [comment]
Hi @muki and @tonekk, did you create Helm Charts in order to deploy the Consul Docker image along with a running postgreSQL service? I'm starting to look into that right now. Maybe it would make sense to build a Consul Operator later on. Thanks.

**Comment by ruslanguns** [comment]
Hello @muki, @tonekk and @muki, I am interested in deploy Consul on a Kubernetes environment. Let me know you got an opportunity to create the Helm Charts or manifest for it. Thanks in advance!

**Comment by javierm** [comment]
Hi, @pwilimzig :smile:.

Since you mentioned in #4872 that tonekk has left the team, do you think you can provide the link to the repo with [the Dockerfile tonekk mentioned](https://github.com/consuldemocracy/consuldemocracy/issues/4797#issuecomment-1225381305)?

**Comment by javierm** [comment]
@muki @Kremer80 @ruslanguns @pwilimzig Does anyone on this issue have experience using [Kamal](https://kamal-deploy.org/)? We're checking whether it'd make sense to use it to install Consul Democracy with Docker on production :thinking:.

**Comment by pwilimzig** [comment]
I have never heard of Kamal. Sorry.

**Comment by javierm** [comment]
@muki @Kremer80 @ruslanguns @pwilimzig In case you know someone who might be interested in applying, there's a [production-ready configuration for Docker project proposal](https://ellak.gr/wiki/index.php?title=Google_Summer_of_Code_2025_proposed_ideas#Docker_for_Consul_Democracy_citizen_participation_platform) for the Google Summer of Code 2025.
